### PR TITLE
Fix git log command arguments and reduce allocations

### LINKED ISSUE #4516: Is there a bug in RevisionGraph.ProccessGitLog?
On current `master` see `RevisionGraph.ProccessGitLog` line 169:

https://github.com/gitextensions/gitextensions/blob/master/GitCommands/RevisionGraph.cs#L169

It appears as though the string assignment should be a string append. That is,

```c#
// before
logParam = " --branches";

// after
logParam += " --branches";
```

If this is intentional, let's add some documentation to that effect.

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #4516.

Changes proposed in this pull request:
 - Fix the git log command argument construction bug
 - Remove temporary strings during argument construction

What did I do to test the code and ensure quality:
 - Tested that string output was the same before/after
 - Manual tests
 - Unit tests

Has been tested on:
 - GIT 2.16
 - Windows 10

### CODE REVIEW

File: GitCommands/RevisionGraph.cs [hunk 1]
```diff
@@ -123,83 +123,71 @@ private void ProccessGitLog(CancellationToken taskState)
...
-            if (AppSettings.OrderRevisionByDate)
-            {
-                logParam = " --date-order";
-            }
-            else
-            {
-                logParam = " --topo-order";
-            }
+            string branchFilter = BranchFilter;
+            if (!string.IsNullOrWhiteSpace(BranchFilter) && BranchFilter.IndexOfAny(ShellGlobCharacters) != -1)
```
Comment (gerhardol) [inline GitCommands/RevisionGraph.cs]: --branches is still incorrectly handled. Move this to where --branches is handled below

Comment (drewnoakes) [reply GitCommands/RevisionGraph.cs]: Thanks. Please check whether 0842600 covers what you were seeing.

**Comment by RussKie** [comment]
Please squash and it's ready to go

**Comment by gerhardol** [comment]
@RussKie 
> Please squash and it's ready to go

Better to squash merge

**Comment by RussKie** [comment]
Not always, as the merger often needs to craft a commit message which is a
rensponsibility of the committer.
Also it is also possible that a PR *may* contain a number of commits
(functional and maintenance), yet some of those need to be squashed.

**Comment by Radrik5** [comment]
This PR broke so powerful Branches filter. Previously I used "@ @{u}" to filter current branch and its upstream branch and "--reflog" to restore commits that lost references. I could also enter several references delimited with space. Now all this doesn't work. Please don't add "--branches=" prefix to the entered string, pass it directly to the git log command line.

**Comment by drewnoakes** [comment]
As commented at your other issue, this PR was merged after 2.51 was released, so I don't see how it could be the cause of the breakage.

**Pull request merged**